### PR TITLE
Update where mysqlreceiver establishes a connection

### LINKED ISSUE #6636: [mysqlreceiver] - Client is only initialized in start function
**Describe the bug**
The mysqlreceiver only initializes a client in its start function. This leaves it vulnerable to transient errors.

**What did you expect to see?**
Client should be initialized in the scrape function.

### PR DESCRIPTION
**Pull request by Mrod1598**
**Description:** <Describe what has changed.>
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6636 through establishing the connection in the scrape func.

Also, Adds a basic integration test to test it.

### CODE REVIEW

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.5_7 [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    apt-get update && \
+    apt-get install -y libmariadb3
+
+COPY scripts/setup.sh /setup.sh
+RUN chmod +x /setup.sh
+
+ENV MYSQL_DATABASE=otel
+ENV MYSQL_USER=otel
+ENV MYSQL_PASSWORD=otel
+ENV MYSQL_ROOT_PASSWORD=otel
```
Comment (djaglowski) [inline receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.5_7:13]:
```suggestion
ENV MYSQL_ROOT_PASSWORD=otel

```

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0 [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    apt-get update && \
+    apt-get install -y libmariadb3
+
+COPY scripts/setup.sh /setup.sh
+RUN chmod +x /setup.sh
+
+ENV MYSQL_DATABASE=otel
+ENV MYSQL_USER=otel
+ENV MYSQL_PASSWORD=otel
+ENV MYSQL_ROOT_PASSWORD=otel
```
Comment (djaglowski) [inline receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0:13]:
```suggestion
ENV MYSQL_ROOT_PASSWORD=otel

```

File: receiver/mysqlreceiver/integration_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
+package mysqlreceiver
```
Comment (djaglowski) [inline receiver/mysqlreceiver/integration_test.go:1]: Needs the copyright notice. Also please add "integration" build tags.

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0 [hunk 4]
```diff
@@ -0,0 +1,13 @@
+FROM mysql:8.0.25
```
Comment (djaglowski) [inline receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0:1]: Can we rename this directory to `mysqlreceiver/testdata/integration`?

File: receiver/mysqlreceiver/integration_test.go (hunk 3 above)
Comment (Mrod1598) [reply receiver/mysqlreceiver/integration_test.go:18]: added

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0 (hunk 2 above)
Comment (Mrod1598) [reply receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0]: added

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.5_7 (hunk 1 above)
Comment (Mrod1598) [reply receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.5_7]: added

File: receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0 (hunk 4 above)
Comment (Mrod1598) [reply receiver/mysqlreceiver/testdata/dockerfiles/Dockerfile.mysql.8_0]: renamed

File: receiver/mysqlreceiver/testdata/integration/scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    result="$?"
+    if setup_permissions; then
+        echo "Permissions configured!"
+        exit 0
+    fi
+    echo "Trying again in 5 seconds. . ."
+    sleep 5
+done
+
+echo "Failed to configure permissions"
```
Comment (djaglowski) [inline receiver/mysqlreceiver/testdata/integration/scripts/setup.sh:29]:
```suggestion
echo "Failed to configure permissions"

```

**Review by djaglowski — COMMENTED** [review]
There are some conflicts based on recent refactoring in #6638. 

Also, we need to update the [list of components with integration tests](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/20f6cead5bcfadb9e82463b2134aa28c777164ac/Makefile#L18) in order for the tests to actually run in Ci.

File: Makefile [hunk 6]
```diff
@@ -23,8 +23,9 @@ INTEGRATION_TEST_MODULES := \
 	receiver/kafkametricsreceiver \
 	receiver/nginxreceiver \
 	receiver/memcachedreceiver \
+	receiver/mysqlreceiver \
 	internal/common \
-	extension/observer/dockerobserver
+	extension/observer/dockerobserver
```
Comment (djaglowski) [inline Makefile:28]: Looks like extra whitespace was unintentionally added here.

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
> A bit unclear if we want this to be the pattern for "Scrapers". Usually these days the clients are "smart" clients that automatically reconnect in case of a failure, is that not the case for SQL client?

I will look into this and determine one way or the other.

Either way, am I correct in understanding that we don't want to fail in `start()` if the client cannot connect? (i.e. Instead, we either rely on a smart client to connect later, or if not smart then recreate a client each time we scrape.)

**Comment by djaglowski** [comment]
@Mrod1598, I made some bad assumptions, detailed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6636#issuecomment-993937669). Thank you @bogdandrutu for pointing this out.

The integration tests that you've added in this PR would still be very valuable additions in my opinion. Would you mind resubmitting those in a new PR? [Here is a tracking issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6804).

**Pull request closed without merging**